src/cmd/keys.js

```diff
@@ -501,16 +502,30 @@ module.exports = class KeysCommand {
 	}
 
 	async _dfuWrite(device, buffer, { altSetting, startAddr, leave, noErase }) {
-		await device.writeOverDfu(buffer, { altSetting, startAddr, leave, noErase });
+		try {
+			await device.writeOverDfu(buffer, { altSetting, startAddr, leave, noErase });
+		} catch (err) {
+			if (err instanceof DeviceProtectionError) {
+				throw new Error('Operation could not be completed due to device protection.');
+			}
+			throw new Error('Writing over DFU failed', err);
+		}
 	}
 
 	async _dfuRead(device, { altSetting, startAddr, size }) {
 		let buf;
 		try {
 			buf = await device.readOverDfu({ altSetting, startAddr, size });
 		} catch (err) {
+			if (err instanceof DeviceProtectionError) {
+				throw new Error('Operation could not be completed due to device protection.');
+			}
 			// FIXME: First time read may fail so we retry
-			buf = await device.readOverDfu({ altSetting, startAddr, size });
+			try {
+				buf = await device.readOverDfu({ altSetting, startAddr, size });
+			} catch (err) {
+				throw new Error('Reading over DFU failed', err);
+			}
```

Comment:
Might as well throw an error here? So I added this

Reply:
That doesn't look necessary. The previous code will throw an error if it fails already.